internal/charmstore/resources_test.go

```diff
@@ -347,63 +371,45 @@ func (s *resourceSuite) TestResolveResource(c *gc.C) {
 			continue
 		}
 		c.Assert(err, jc.ErrorIsNil)
-		adjustExpectedResource(res, resources[test.expectResource])
-		c.Assert(res, jc.DeepEquals, resources[test.expectResource])
+		checkResourceDocs(c, store, id, []string{fmt.Sprintf("someResource/%d", test.expectResource)}, []*mongodoc.Resource{res})
 	}
 }
 
-var publishResourceErrorTests = []struct {
-	about       string
-	channel     params.Channel
-	resources   []mongodoc.ResourceRevision
-	expectError string
-}{{
-	about:       "no channel",
-	channel:     params.NoChannel,
-	expectError: `missing channel`,
-}, {
-	about:       "unpublished channel",
-	channel:     params.UnpublishedChannel,
-	expectError: `cannot publish to unpublished channel`,
-}, {
-	about:   "unknown resource name",
-	channel: params.StableChannel,
-	resources: []mongodoc.ResourceRevision{{
-		Name:     "for-install",
-		Revision: 0,
-	}, {
-		Name:     "bad-name",
-		Revision: 0,
-	}},
-	expectError: `charm does not have resource "bad-name"`,
-}, {
-	about:   "no such resource",
-	channel: params.StableChannel,
-	resources: []mongodoc.ResourceRevision{{
-		Name:     "for-install",
-		Revision: 0,
-	}},
-	expectError: `cs:~charmers/xenial/starsay-3 resource for-install revison 0 not found`,
-}}
+func (s *resourceSuite) TestPublishWithResourceNotInMetadata(c *gc.C) {
+	store := s.newStore(c, false)
+	defer store.Close()
+
+	id := MustParseResolvedURL("cs:~charmers/precise/wordpress-3")
+	err := store.AddCharmWithArchive(id, storetesting.NewCharm(nil))
+	c.Assert(err, jc.ErrorIsNil)
 
-func (s *resourceSuite) TestPublishResourceErrors(c *gc.C) {
+	err = store.Publish(id, map[string]int{
+		"resource1": 0,
+	}, params.StableChannel)
+	c.Assert(err, gc.ErrorMatches, `charm published with incorrect resources: charm does not have resource "resource1"`)
```

Comment:
I wonder if this needs to be cleverer. Specifying N resources, where one resource is not defined, the publish should fail without uploading of any resource, so that the result is always consistent. That means, if resource 1 <= R <=N does not exist, first R-1 resources are not uploaded and only then error given.